sdks/python/hatchet_sdk/clients/v1/api_client.py

```diff
@@ -0,0 +1,89 @@
+import asyncio
+from concurrent.futures import ThreadPoolExecutor
+from typing import AsyncContextManager, Callable, Coroutine, ParamSpec, TypeVar
+from uuid import UUID
+
+from hatchet_sdk.clients.rest.api_client import ApiClient
+from hatchet_sdk.clients.rest.configuration import Configuration
+from hatchet_sdk.config import ClientConfig
+
+## Type variables to use with coroutines.
+## See https://stackoverflow.com/questions/73240620/the-right-way-to-type-hint-a-coroutine-function
+## Return type
+R = TypeVar("R")
+
+## Yield type
+Y = TypeVar("Y")
+
+## Send type
+S = TypeVar("S")
+
+P = ParamSpec("P")
+
+
+def maybe_additional_metadata_to_kv(
+    additional_metadata: dict[str, str] | None
+) -> list[str] | None:
+    if not additional_metadata:
+        return None
+
+    return [f"{k}:{v}" for k, v in additional_metadata.items()]
+
+
+def maybe_uuid_to_str(value: UUID | None) -> str | None:
+    return str(value) if value else None
+
+
+def maybe_uuid_list_to_str_list(value: list[UUID] | None) -> list[str] | None:
+    return [str(uuid) for uuid in value] if value else None
+
+
+class BaseRestClient:
+    def __init__(self, config: ClientConfig) -> None:
+        self.tenant_id = config.tenant_id
+
+        self.client_config = config
+        self.api_config = Configuration(
+            host=config.server_url,
+            access_token=config.token,
+        )
+
+        self.api_config.datetime_format = "%Y-%m-%dT%H:%M:%S.%fZ"
+
+    def client(self) -> AsyncContextManager[ApiClient]:
+        return ApiClient(self.api_config)
```

Comment:
There were also some issues with aiohttp clients not being closed, which I think was happening because we're supposed to be using this as an async context manager, which is what we'll do now